tpm2/src/util/ski.c

```diff
@@ -0,0 +1,560 @@
+      || encodeBase64Data(sk_pub_data, sk_pub_size, &sk64_pub_data,
+                          &sk64_pub_size)
+      || encodeBase64Data(sk_priv_data, sk_priv_size, &sk64_priv_data,
+                          &sk64_priv_size)
+      || encodeBase64Data(wk_pub_data, wk_pub_size, &wk64_pub_data,
+                          &wk64_pub_size)
+      || encodeBase64Data(wk_priv_data, wk_priv_size, &wk64_priv_data,
+                          &wk64_priv_size)
+      || encodeBase64Data(input.enc_data, input.enc_data_size, &enc64_data,
+                          &enc64_data_size))
+  {
+    kmyth_log(LOG_ERR, "error base64 encoding ski string ... exiting");
+    free(pcr_select_data);
+    free(sk_pub_data);
+    free(sk_priv_data);
+    free(wk_pub_data);
+    free(wk_priv_data);
+    free(pcr64_select_data);
+    free(sk64_pub_data);
+    free(sk64_priv_data);
+    free(wk64_pub_data);
+    free(wk64_priv_data);
+    free(enc64_data);
+    return 1;
+  }
+
+  free(pcr_select_data);
+  free(sk_pub_data);
+  free(sk_priv_data);
+  free(wk_pub_data);
+  free(wk_priv_data);
+
+  //At this point the data is all formatted, it's time to create the string
+
+  uint8_t *out = NULL;
+  size_t out_length = 0;
+
+  concat(&out, &out_length, (uint8_t *) KMYTH_DELIM_PCR_SELECTION_LIST,
+         strlen(KMYTH_DELIM_PCR_SELECTION_LIST));
+  concat(&out, &out_length, pcr64_select_data, pcr64_select_size);
+  free(pcr64_select_data);
+
+  concat(&out, &out_length, (uint8_t *) KMYTH_DELIM_STORAGE_KEY_PUBLIC,
+         strlen(KMYTH_DELIM_STORAGE_KEY_PUBLIC));
+  concat(&out, &out_length, sk64_pub_data, sk64_pub_size);
+  free(sk64_pub_data);
+
+  concat(&out, &out_length, (uint8_t *) KMYTH_DELIM_STORAGE_KEY_PRIVATE,
+         strlen(KMYTH_DELIM_STORAGE_KEY_PRIVATE));
+  concat(&out, &out_length, sk64_priv_data, sk64_priv_size);
+  free(sk64_priv_data);
+
+  concat(&out, &out_length, (uint8_t *) KMYTH_DELIM_CIPHER_SUITE,
+         strlen(KMYTH_DELIM_CIPHER_SUITE));
+  concat(&out, &out_length, (uint8_t *) input.cipher.cipher_name,
+         strlen(input.cipher.cipher_name));
+  concat(&out, &out_length, (uint8_t *) "\n", 1);
+
+  concat(&out, &out_length, (uint8_t *) KMYTH_DELIM_SYM_KEY_PUBLIC,
+         strlen(KMYTH_DELIM_SYM_KEY_PUBLIC));
+  concat(&out, &out_length, wk64_pub_data, wk64_pub_size);
+  free(wk64_pub_data);
+
+  concat(&out, &out_length, (uint8_t *) KMYTH_DELIM_SYM_KEY_PRIVATE,
+         strlen(KMYTH_DELIM_SYM_KEY_PRIVATE));
+  concat(&out, &out_length, wk64_priv_data, wk64_priv_size);
+  free(wk64_priv_data);
+
+  concat(&out, &out_length, (uint8_t *) KMYTH_DELIM_ENC_DATA,
+         strlen(KMYTH_DELIM_ENC_DATA));
+  concat(&out, &out_length, enc64_data, enc64_data_size);
+  free(enc64_data);
+
+  concat(&out, &out_length, (uint8_t *) KMYTH_DELIM_END_FILE,
+         strlen(KMYTH_DELIM_END_FILE));
+
+  *output = out;
+  *output_length = out_length;
+
+  return 0;
+}
+
+void free_ski(Ski * ski)
+{
+  free(ski->enc_data);
```

Comment:
Dow we not have to `free` the TPM structures in the `Ski`?

Reply:
I don't think there are any other internal pointers needing to be freed.